# General data questions and exploration


In [2]:
### Initialisation ###
import os
import pandas as pd
import matplotlib.pyplot as plt

%matplotlib inline
plt.rcParams['figure.figsize'] = [6, 4]

EVENT_NAME = Event.objects.all()[0].name.replace(' ', '')
DIR = './data/harvey_user_location/'
DF_FILENAME = 'df_users.csv'
LATEX_DAT_PATH = DIR + 'latex/' + 'harvey-exploration.dat'
LATEX_TABLE_PATH = DIR + 'latex/' + 'harvey-exploration-tabs.dat'

# Confirm correct database is set in Django settings.py
if 'Harvey' not in EVENT_NAME:
    raise Exception('Event name mismatch -- check database set in Django')

for file in [LATEX_DAT_PATH, LATEX_TABLE_PATH]:
    try:    
        os.remove(file)
    except:
        pass

# Open original Dataframe
users_df = pd.read_csv(DIR + DF_FILENAME, index_col=0)
users_df.shape

(1500, 46)

In [3]:
def set_tex_var(name, val, dec=3):
    """
    Define a value as a latex variable to be
    imported into Latex file with (e.g.):
    \input{data/harvey-coding.dat}

    NOTE: does not look for duplicate variable definitions.
    """
    if isinstance(val, float):
        template = r'\newcommand{{\{}}}{{{:,.' + str(dec) + r'f}}}'
    elif isinstance(val, int):
        template = r'\newcommand{{\{}}}{{{:,}}}'
    else:
        template = r'\newcommand{{\{}}}{{{}}}'
    with open(LATEX_DAT_PATH, 'a') as a_writer:
        a_writer.write(template.format(name, val) + '\n')

In [4]:
def sanitise_tex_str(string):
    '''
    Sanitises string for use in Latex
    NOTE: Incomplete
    '''
    return string.replace('%', '\%')


def df_to_tex(df, cols=None, label='UNNAMED', caption='UNDEFINED', 
              width=0.8, row_sep=1, col_widths=None, dec_dict=None):
    '''
    Generates a latex table from a dataframe, 
    to be copied by user from output file to tex
    
    NOTE: format columns in the dataframe before passing to func, eg:
        df[col] = df[col].astype(int).map('{:.1f}%'.format)
        df.rename(columns={"A": "a", "B": "c"}, inplace=True)
    NOTE: linewidth and column widths are fixed and require manual adjustment
    
    input: dataframe and desired columns, 
            table reference name and caption text
    output: appends to file LATEX_TABLE_PATH
    returns: none
    '''
    df = df.copy()
    if not cols:
        cols = df.columns
    for col in [x for x in cols if x not in df.columns]:
        df[col] = ''
    df = df[cols]
    
    print('Writing df to file as tex...')
    with open(LATEX_TABLE_PATH, 'a') as a_writer:
        a_writer.write('%<*{}>'.format(label) + '\n')
#        a_writer.write('tab:' + label + '\n\n')
        a_writer.write('\\begin{table}[H]' + '\n')
        a_writer.write('  \\small\\vspace{5mm}\\centering\\extrarowsep=' + str(row_sep) + 'mm' + '\n')
        if not col_widths:
            col_widths = 'X[-1,lm] ' + ('X[-2,lm] ' * (len(df.columns)-1))
        a_writer.write('  \\begin{tabu} to ' + str(width) + '\\linewidth { '+ col_widths + '}' + '\n')
        a_writer.write('    \\toprule\n')
        
        row_string = '    '
        for col in cols:
            # If last column, add line ending
            if col == df.columns[-1]:
                row_string += col + ' \\\\\n'
            else:
                # Pad to length of longest column value, plus a fixed amount:
                pad = max(df[col].map(str).map(len)) + 4
                if pad - len(col) > 18 or len(col) > pad: # (ignore pad for titles if col vals are long)
                    row_string += (sanitise_tex_str(col) + ' & ')
                else:
                    row_string += (sanitise_tex_str(col).ljust(pad) + '& ')
#                sum([len(x) for x in cols]) + (len(cols)*3) # Ignore pad if col titles exceed text display width ?
        a_writer.write(row_string)
        
        a_writer.write('    \\midrule\n')
        
        for index, row in df.iterrows():
            row_string = '    '
            for col in df.columns:
                if col == df.columns[-1]:
                    if dec_dict and col in dec_dict.keys():
                        row_string += (sanitise_tex_str(
                            ('{:.' + str(dec_dict[col]) + 'f}').format(row[col]) 
                        ) + ' \\\\\n')
                    else:
                        row_string += (sanitise_tex_str(str(row[col])) + ' \\\\\n')
                else:
                    pad = max(df[col].map(str).map(len)) + 4
                    if dec_dict and col in dec_dict.keys(): #dec_dict is used to force decimal places with trailing zeroes
                        row_string += (sanitise_tex_str(
                            ('{:.' + str(dec_dict[col]) + 'f}').format(row[col]) 
                        ).ljust(pad) + '& ')
                    else:
                        row_string += (sanitise_tex_str(str(row[col])).ljust(pad) + '& ')
            a_writer.write(row_string)
        
        a_writer.write('    \\bottomrule' + '\n')
        
        a_writer.write('  \\end{tabu}' + '\n')
        a_writer.write('  \\caption[{}]{{{}}}'.format(caption, caption) + '\n')
        a_writer.write('  \\label{{tab:{}}}'.format(label) + '\n')
        a_writer.write('\\end{table}' + '\n')
        a_writer.write('%</{}>'.format(label) + '\n\n\n')

        return


In [8]:
import pandas as pd
import math
import numpy as np
#from sklearn.metrics import matthews_corrcoef


def confusion_matrix(df: pd.DataFrame, col1: str, col2: str):
    """
    Given a dataframe with at least
    two categorical columns, create a 
    confusion matrix of the count of the columns
    cross-counts
    """
    return (
            df
            .groupby([col1, col2])
            .size()
            .unstack(fill_value=0)
            )


def confusion_matrix_from_series(s1, s2):
    """
    Returns confusion matrix for two binary
    series
    """
    df = pd.concat([s1, s2], axis=1)
    return confusion_matrix(df, s1.name, s2.name)


def calc_agreement_coefs(df: pd.DataFrame):
    """
    Calculates Cohen's Kappa and
    Krippendorff's Alpha for a
    given confusion matrix.
    """
    arr = df.to_numpy()
    n = arr.sum()
    p_o = 0
    for i in range(len(arr)):
        p_o += arr[i][i]/n
    p_e = 0
    for i in range(len(arr)):
        p_e += (arr.sum(axis=1)[i] *
                arr.sum(axis=0)[i]) / (n*n)
    kappa = (p_o-p_e)/(1-p_e)
    
    phi_num = (arr[0][0]*arr[1][1]) - (arr[0][1]*arr[1][0])
    a = np.sum(conf, axis=1)
    b = np.sum(conf, axis=0)
    phi_denom = math.sqrt(np.prod([a, b]))
    phi = phi_num/phi_denom
    phi_max = min(
        math.sqrt(a[0] * b[1])/math.sqrt(a[1] * b[0]),
        math.sqrt(a[1] * b[0])/math.sqrt(a[0] * b[1])
        )
    phiphimax = phi/phi_max
    
    coin_arr = np.transpose(arr) + arr
    exp_distribution = [sum(x) for x in coin_arr]
    p_e_krippendorf = sum([a * (a-1) for a in exp_distribution])/(2*n*((2*n)-1))
    alpha = (p_o - p_e_krippendorf) / (1-p_e_krippendorf)
    
    return p_o, kappa, alpha, phi, phiphimax


def calc_agreement_metrics(df: pd.DataFrame):
    """
    Calculates various agreement metrics
    for a given binary confusion matrix.
    
    Assumes true condition as ROW heading and
    ascending integer labels.
    """
    arr = df.to_numpy()
    if len(arr) != 2:
        print('ERROR: df is not 2x2')
        return None
    results = {}
    results['Prevalence'] = arr.sum(axis=0)[1]/arr.sum()
    results['Accuracy'] = (arr[0][0] + arr[1][1])/arr.sum()
    results['Prec'] = arr[1][1]/arr.sum(axis=1)[1]
    results['Recall'] = arr[1][1]/arr.sum(axis=0)[1]
    results['f1Score'] = (2 * results['Prec'] * results['Recall'])/(results['Prec']+results['Recall'])
    results['Specificity'] = arr[0][0]/arr.sum(axis=0)[0]
    results['FalseNegRate'] = arr[0][1]/arr.sum(axis=0)[1]
    p_o, kappa, alpha, phi, phiphimax = calc_agreement_coefs(df)
    results['Cohen\'s Kappa'] = kappa
    results['Krippendorff\'s Alpha'] = alpha
    results['Phi Coef / MCC'] = phi
    results['Phi/PhiMax'] = phiphimax
    
    return results
    

First, given we are treating the data as binary-coded, we need to be aware of the third code of 'unsure'. In most comparisons, these will automatically be considered as the opposite code from what is being tested. We can instead manually set it to either value, or remove it entirely, given that it comprises a small proportion of the data.

In [5]:
unsure_code = (users_df.is_coded_as_witness == 0) & (users_df.is_coded_as_non_witness == 0)
print(sum(unsure_code), 'cases coded as \'unsure\'')

# Remove 'unsure' rows from data:
#users_df = users_df.loc[unsure_code==False]

# Assign 'unsure' rows to positive coded case:
#users_df.is_coded_as_witness = (users_df.is_coded_as_non_witness == False).astype(int)

# Assign 'unsure' rows to negative coded case:
#users_df.is_coded_as_non_witness = (users_df.is_coded_as_witness == False).astype(int)

31 cases coded as 'unsure'


## Geographic Metadata and Manual Coding
Manual coding of users targetted the perceived locality of the user to the event. We can compare the geographic metadata provided by Twitter to these codes to determine their usefulness as a predictor for this value.

### Profile Location Field
The first value to check is the location of a user as set in their profile. This is a user-set string. In an earlier notebook, this string was geocoded using Google Maps api and evaluated for whether it fell within the bounding box defined for this event. We can therefore check whether this test correlates with the coded value.

First, we check what proportion of users provide a value in the field. We can then generate a confusion matrix showing the agreement between the profile locality where provided, and the coded value.

In [6]:
us = User.objects.filter(user_class__gt=0)
#us = us.filter(coding_for_user__coding_id=1, coding_for_user__data_code__data_code_id__gt=0)
tot = us.count()
tot_loc = us.count() - (us.filter(location="") | us.filter(location__isnull=True)).count()

print('Total users: ', tot)
print('Total users with location filled: ', tot_loc)
print('Proportion: {:.4}%'.format((tot_loc/tot)*100))

print('\nProportion of coded users with location filled: {:.4}%'.format((sum(users_df.location.notna())/users_df.shape[0])*100))

print('\nProportion of coded users with parseable location filled: {:.4}%'.format(
    (users_df.loc[(users_df.is_non_local_profile_location + users_df.is_local_profile_location) > 0].shape[0]/users_df.shape[0])*100))


Total users:  31932
Total users with location filled:  25619
Proportion: 80.23%

Proportion of coded users with location filled: 79.4%

Proportion of coded users with parseable location filled: 77.47%


In [7]:
vals = users_df.loc[users_df["is_local_profile_location"] == 1]["is_coded_as_witness"].value_counts()
vals2 = users_df.loc[users_df["is_coded_as_witness"] == 1]["is_local_profile_location"].value_counts()

print('{} of {} ({:.4}%) users were classified as having a local profile'.format(sum(vals), len(users_df), sum(vals)/len(users_df)*100))
print('{} of {} ({:.4}%) users were coded as a witness'.format(sum(vals2), len(users_df), sum(vals2)/len(users_df)*100))
print('{} of {} ({:.4}%) users with local profile locations were coded as witness'.format(vals[1], sum(vals), vals[1]/sum(vals)*100))
print('{} of {} ({:.4}%) witness codes had a local profile'.format(vals2[1], sum(vals2), vals2[1]/sum(vals2)*100))

397 of 1500 (26.47%) users were classified as having a local profile
386 of 1500 (25.73%) users were coded as a witness
258 of 397 (64.99%) users with local profile locations were coded as witness
258 of 386 (66.84%) witness codes had a local profile


In [9]:
# We exclude rows where either no profile location field was provided, 
# or the location was not parsed by the API:
loc_df = users_df.loc[(users_df.is_non_local_profile_location + users_df.is_local_profile_location) > 0]
print(users_df.shape[0] - loc_df.shape[0], 'rows with no parseable profile location value excluded')

conf = confusion_matrix(loc_df, 'is_local_profile_location', 'is_coded_as_witness')
conf

338 rows with no parseable profile location value excluded


is_coded_as_witness          0    1
is_local_profile_location          
0                          695   70
1                          139  258

In [10]:
results = calc_agreement_metrics(conf)
res_df = pd.DataFrame.from_dict(results, orient='index', columns=['loc_prof_nona'])
res_df

loc_prof_nona
Prevalence                 0.282272
Accuracy                   0.820138
Prec                       0.649874
Recall                     0.786585
f1Score                    0.711724
Specificity                0.833333
FalseNegRate               0.213415
Cohen's Kappa              0.582731
Krippendorff's Alpha       0.581198
Phi Coef / MCC             0.588340
Phi/PhiMax                 0.675833

As excluding the ~20% of values with no parseable location field provided is not an option in practice, we must decide to either discard them (i.e. by default classify as non-local) or include them (default classify as local). The first option will inevitable discard true positive cases, thus reducing recall, whereas the latter will introduce false positives, reducing precision:

In [11]:
conf = confusion_matrix(users_df, 'is_local_profile_location', 'is_coded_as_witness')
print('loc_prof_notna')
print(conf)
res_df['loc_prof_notna'] = calc_agreement_metrics(conf).values()


conf = confusion_matrix_from_series(
    pd.Series(
                (users_df.is_local_profile_location) | (users_df.location.isna()).astype(int), 
                name='is_local_profile_location_or_na'
             ),
    users_df.is_coded_as_witness)
print('\nloc_prof_orna')
print(conf)
res_df['loc_prof_orna'] = calc_agreement_metrics(conf).values()


res_df

loc_prof_notna
is_coded_as_witness          0    1
is_local_profile_location          
0                          975  128
1                          139  258

loc_prof_orna
is_coded_as_witness                0    1
is_local_profile_location_or_na          
0                                714   80
1                                400  306


loc_prof_nona  loc_prof_notna  loc_prof_orna
Prevalence                 0.282272        0.257333       0.257333
Accuracy                   0.820138        0.822000       0.680000
Prec                       0.649874        0.649874       0.433428
Recall                     0.786585        0.668394       0.792746
f1Score                    0.711724        0.659004       0.560440
Specificity                0.833333        0.875224       0.640934
FalseNegRate               0.213415        0.331606       0.207254
Cohen's Kappa              0.582731        0.538603       0.341243
Krippendorff's Alpha       0.581198        0.538725       0.309098
Phi Coef / MCC             0.588340        0.538700       0.379833
Phi/PhiMax                 0.675833        0.549040       0.608462

The results above are as expected. Excluding empty fields gives a precision/recall of 0.650/0.668 whereas exluding them gives 0.433/0.793. The exclusion strategy provides the highest f1 score, however given the purpose of the algorithm must be considered when choosing how to weight precision and recall metrics. For example, given the algorithm is designed to curate the feed for human consumption, a high precision is only necessary if the rate of positive cases exceeds the humans' ability to parse the incoming stream. Where the rate is low, sacrificing precision is acceptable to present the human user with more cases which they can then manually filter. 

This concept will be explored in more depth later in the project. For now, it is sufficient to note the values as a baseline model.

### Tweet Stream Coordinates
When posting a Tweet, a user may attach geographic coordinates. The location of the device is provided by the hardware and automatically included with the Tweet (thus the user does not influence the input). A Tweet may also include, instead of specific coordinates, a 'Place' object -- a geographic region (defined by Twitter) which typically describes a location such as a city, state or other similarly-sized region.

To geolocate a user, we can therefore investigate their Twitter feed for any Tweets containing this geographic data and compare these to the bounding box of the observed event. The derived field therefore represents whether *any* of a user's Tweets were identified as 'local' during the event.

For this dataset, the Twitter feed for each observed user spanning the duration of the collection period was collected at the end of the collection period. The feed is therefore made up of Tweets detected during the collection period, and any other Tweets the user made during the period, before or after the detected Tweet, provided they existed at the end of the collection period.

Further work on this area should consider the following:
* Where a local Tweet has been detected, check the proportion of other Tweets containing geographic data.
* Where other geo-Tweets exist, check whether they are from the same point or move around -- consider recoding for where a user Tweets from within *and* without the bounding box.
* Where all geo-Tweets come from the same point, it is likely that location has been manually set (i.e. it is a storefront/business account). This may be verifiable by checking the Tweet source.

In [12]:
res_df['local_tw'] = calc_agreement_metrics(conf).values()
res_df

loc_prof_nona  loc_prof_notna  loc_prof_orna  local_tw
Prevalence                 0.282272        0.257333       0.257333  0.257333
Accuracy                   0.820138        0.822000       0.680000  0.680000
Prec                       0.649874        0.649874       0.433428  0.433428
Recall                     0.786585        0.668394       0.792746  0.792746
f1Score                    0.711724        0.659004       0.560440  0.560440
Specificity                0.833333        0.875224       0.640934  0.640934
FalseNegRate               0.213415        0.331606       0.207254  0.207254
Cohen's Kappa              0.582731        0.538603       0.341243  0.341243
Krippendorff's Alpha       0.581198        0.538725       0.309098  0.309098
Phi Coef / MCC             0.588340        0.538700       0.379833  0.379833
Phi/PhiMax                 0.675833        0.549040       0.608462  0.608462

While the recall of this metric is close (slightly less than) `loc_prof_notna`, there is no increase in precision and thus it is an inferior metric to predict the true condition.

We can however check whether the metric is capturing a different proportion of local users, and therefore improve upon the existing measures through combination. Using an OR condition will increase recall at the cost of precision; using an AND condition will increase precision at the cost of recall.

In [13]:
conf = confusion_matrix(users_df, 'has_tweet_from_locality', 'is_coded_as_witness')
conf

is_coded_as_witness        0    1
has_tweet_from_locality          
0                        898  135
1                        216  251

In [14]:
conf = confusion_matrix_from_series(
    pd.Series(
        ((users_df.is_local_profile_location) | (users_df.has_tweet_from_locality)),
        name = 'is_local_profile_location_or_local_tw'
        ),
    users_df.is_coded_as_witness)
# conf

res_df['loc_prof_notna_or_loc_tw'] = calc_agreement_metrics(conf).values()

conf = confusion_matrix_from_series(
    pd.Series(
        ((users_df.is_local_profile_location) & (users_df.has_tweet_from_locality)),
        name = 'is_local_profile_location_and_local_tw'
        ),
    users_df.is_coded_as_witness)
# conf

res_df['loc_prof_notna_and_loc_tw'] = calc_agreement_metrics(conf).values()

res_df

loc_prof_nona  loc_prof_notna  loc_prof_orna  local_tw  \
Prevalence                 0.282272        0.257333       0.257333  0.257333   
Accuracy                   0.820138        0.822000       0.680000  0.680000   
Prec                       0.649874        0.649874       0.433428  0.433428   
Recall                     0.786585        0.668394       0.792746  0.792746   
f1Score                    0.711724        0.659004       0.560440  0.560440   
Specificity                0.833333        0.875224       0.640934  0.640934   
FalseNegRate               0.213415        0.331606       0.207254  0.207254   
Cohen's Kappa              0.582731        0.538603       0.341243  0.341243   
Krippendorff's Alpha       0.581198        0.538725       0.309098  0.309098   
Phi Coef / MCC             0.588340        0.538700       0.379833  0.379833   
Phi/PhiMax                 0.675833        0.549040       0.608462  0.608462   

                      loc_prof_notna_or_loc_tw  loc_prof_notna_and_loc_tw  
Prevalence                            0.257333                   0.257333  
Accuracy                              0.788000                   0.800000  
Prec                                  0.558219                   0.653571  
Recall                                0.844560                   0.474093  
f1Score                               0.672165                   0.549550  
Specificity                           0.768402                   0.912926  
FalseNegRate                          0.155440                   0.525907  
Cohen's Kappa                         0.524972                   0.425170  
Krippendorff's Alpha                  0.515676                   0.421208  
Phi Coef / MCC                        0.549560                   0.434221  
Phi/PhiMax                            0.745458                   0.533534

As we can see from the table above, the highest precision is observed from the classifier `loc_prof_notna_and_loc_tw` at 0.656 (though at a great cost to recall). It should be noted that selecting for cases which satisfy both conditions may inadvertently select for a particular (as-yet unidentified) sub-category of user type and exclude other categories of value. The highest fscore comes from `loc_prof_notna_or_loc_tw` at 0.672, which has a precision of 0.558.

While the low scores of these metrics show that they cannot provide meaningful proxies for the manually assigned code, as classifiers they provide a suitable baseline from which to measure more sophisticated models.

## Tweet Source
An important distinguishing metadatum of a Tweet is the 'source' field, which represents the platform from which the Tweet was published. When creating a third-party application which can interact with the Twitter API, a developer must provide a descriptor string which populates this field. Because many third-party applications are designed for specific use-cases, this field provides useful information which characterises the motivations for conditions under which the Tweet was created. For example, the source `TweetMyJOBS` refers to a recruitment platform and thus is attached to Tweets advertising job listings.

We can look at the list of most common source from within the entire dataset and compare this to the sources detected during the collection period. (Note that the complete dataset will still contain a selection bias and does not necessarily characterise regular Twitter use)

In [15]:
# View most common sources from entire dataset:
from django.db.models import Count
from streamcollect.models import Tweet

ts = Tweet.objects.all()

fieldname = 'source'
counts = ts.values(fieldname).order_by(fieldname).annotate(count=Count(fieldname)).order_by('-count')

dft = pd.DataFrame.from_dict(counts[:10])
dft['Proportion'] = (dft['count'].astype(int) / ts.count()*100).map('{:.1f}%'.format)

# Title case for columns:
dft.rename(columns={x: x.title() for x in dft.columns}, inplace=True)

print('Total Tweets:', ts.count())
dft

Total Tweets: 1727438


Count               Source Proportion
0  535558   Twitter for iPhone      31.0%
1  348743  Twitter for Android      20.2%
2  342908   Twitter Web Client      19.9%
3   95360                IFTTT       5.5%
4   45830     Twitter for iPad       2.7%
5   43808         Twitter Lite       2.5%
6   42045            Instagram       2.4%
7   32273            TweetDeck       1.9%
8   28536             Facebook       1.7%
9   27105            Hootsuite       1.6%

In [16]:
# Adding descriptions for latex table, based on manual inspection and classification:
descs = [
    'first-party', 'first-party', 'first-party', 'automated process, crossposting', 'first-party',
    'first-party', 'third-party crossposting', 'first-party', 'third-party crossposting', 'social media manager'
]
dft['Description'] = pd.Series(descs)

df_to_tex(dft, cols=['Proportion', 'Source', 'Description'], label='top-tw-sources', caption='Top 10 Tweet Sources')

Writing df to file as tex...


In [17]:
# Count proportion of Tweets from first-party applications
first_party_sources = [
    'Twitter for iPhone',
    'Twitter for Android',
    'Twitter Web Client',
    'Twitter for iPad',
    'Twitter Lite',
    'TweetDeck',
    'Twitter for Windows',
    'Twitter for Mac',
    'Twitter for Windows Phone',
    'Twitter for BlackBerry',
    'Twitter for Android Tablets',
    'Twitter MMS'
    ]
fp_count = 0
for x in counts:
    if any([y in x['source'] for y in first_party_sources ]):
        fp_count += x['count']
print('{:.1f}% of total from first party clients: {}'.format((fp_count/ts.count()*100), fp_count))

78.5% of total from first party clients: 1355569


In [18]:
# Write to Latex variable
set_tex_var('hrvTwTotal', ts.count())
set_tex_var('hrvTwFP', fp_count)
set_tex_var('hrvTwFPPrp', fp_count/ts.count()*100, 1)

In [19]:
ts = Tweet.objects.filter(data_source=1)
print('Total for data_source=1 (keyword stream):', ts.count())

fieldname = 'source'
counts = ts.values(fieldname).order_by(fieldname).annotate(count=Count(fieldname)).order_by('-count')

dft = pd.DataFrame.from_dict(counts[:10])
dft['Proportion'] = (dft['count'].astype(int) / ts.count()*100).map('{:.1f}%'.format)

# Title case for columns:
dft.rename(columns={x: x.title() for x in dft.columns}, inplace=True)

dft

Total for data_source=1 (keyword stream): 31303


Count               Source Proportion
0   8959   Twitter for iPhone      28.6%
1   7989   Twitter Web Client      25.5%
2   5086  Twitter for Android      16.2%
3   1754             Paper.li       5.6%
4    981            Hootsuite       3.1%
5    733            Instagram       2.3%
6    720                IFTTT       2.3%
7    682             Facebook       2.2%
8    658            TweetDeck       2.1%
9    652     Twitter for iPad       2.1%

In [20]:
ts = Tweet.objects.filter(data_source__gt=1, coordinates_lat__isnull=False)
print('Total for data_source=3 (geo stream)', ts.count())

fieldname = 'source'
counts = ts.values(fieldname).order_by(fieldname).annotate(count=Count(fieldname)).order_by('-count')

dft = pd.DataFrame.from_dict(counts[:10])
dft['Proportion'] = (dft['count'].astype(int) / ts.count()*100).map('{:.1f}%'.format)

# Title case for columns:
dft.rename(columns={x: x.title() for x in dft.columns}, inplace=True)

dft

Total for data_source=3 (geo stream) 15569


Count                    Source Proportion
0  11922                 Instagram      76.6%
1   1020               TweetMyJOBS       6.6%
2    928  SafeTweet by TweetMyJOBS       6.0%
3    602                BubbleLife       3.9%
4    267                Foursquare       1.7%
5    234                   Untappd       1.5%
6    129       Twitter for Android       0.8%
7    104                 Hootsuite       0.7%
8     87        Twitter for iPhone       0.6%
9     55                 circlepix       0.4%

In [21]:
# Adding descriptions for latex table, based on manual inspection and classification:
descs = [
    'third-party crossposting', 'job listings', 'job listings', 'social media manager', 'check-in app, crossposting',
    'check-in app, crossposting', 'first-party', 'social media manager', 'first-party', 'real estate listings'
    ]
dft['Description'] = pd.Series(descs)

df_to_tex(dft, cols=['Proportion', 'Source', 'Description'], label='top-tw-sources-geo', caption='Top 10 Tweet Sources -- Geographic Stream')

Writing df to file as tex...


In [22]:
fp_count = 0
for x in counts:
    if any([y in x['source'] for y in first_party_sources ]):
        fp_count += x['count']
print('{:.1f}% of geo stream from first party clients: {}'.format((fp_count/ts.count()*100), fp_count))

1.7% of geo stream from first party clients: 259


In [23]:
set_tex_var('hrvTwGeoTotal', ts.count())
set_tex_var('hrvTwGeoFP', fp_count)
set_tex_var('hrvTwGeoFPPrp', fp_count/ts.count()*100, 1)

val = ts.filter(source='Instagram').count()
set_tex_var('hrvTwGeoInsta', val)
set_tex_var('hrvTwGeoInstaPrp', val/ts.count()*100, 1)

val = ts.filter(source__contains='TweetMyJOBS').count()
set_tex_var('hrvTwGeoJobs', val)
set_tex_var('hrvTwGeoJobsPrp', val/ts.count()*100, 1)

Of the entire dataset of 1,727,438 Tweets, those published by first-party Twitter clients comprised 78.5% (1,355,569). In contrast, of the subset of 15,630 Tweets collected based on their location within the event's bounding box, only 1.7% (262) were published from first-party apps. Tweets crossposted by Instagram comprised 76.6% of the geotagged Tweets. The high incidence of Instagram posts in the geographic stream therefore suggest that Instagram posts are much more likely than Tweets to include geographic data, which is preserved during the crossposting process. We can check the rate across the entire dataset excluding the geo stream:

In [24]:
ts = Tweet.objects.filter(source='Instagram', data_source__lt=3)
print('Total Instagram Tweets (excl. geo stream):', ts.count())
ts_geo = ts.filter(coordinates_lat__isnull=False)
print('Total geotagged Instagram Tweets: {}, {:.1f}%'.format(ts_geo.count(), ts_geo.count()/ts.count()*100 ))

Total Instagram Tweets (excl. geo stream): 30123
Total geotagged Instagram Tweets: 9818, 32.6%


In [25]:
set_tex_var('hrvTwInstaTotalNongeo', ts.count())
set_tex_var('hrvTwInstaGeotagNongeo', ts_geo.count())
set_tex_var('hrvTwInstaGeotagNongeoPrp', ts_geo.count()/ts.count()*100, 1)

In [26]:
ts = Tweet.objects.filter(source__contains='TweetMyJOBS')
print('Total TweetMyJOBS Tweets:', ts.count())
ts_geo = ts.filter(coordinates_lat__isnull=False)
print('Total geotagged TweetMyJOBS Tweets: {}, {:.1f}%'.format(ts_geo.count(), ts_geo.count()/ts.count()*100 ))

Total TweetMyJOBS Tweets: 18517
Total geotagged TweetMyJOBS Tweets: 17150, 92.6%


In [27]:
set_tex_var('hrvTwJobsTotal', ts.count())
set_tex_var('hrvTwJobsGeotag', ts_geo.count())
set_tex_var('hrvTwJobsGeotagPrp', ts_geo.count()/ts.count()*100, 1)

As Instagram posts are traditionally based upon the publication of a recently-taken photo, the high incidence of geotagging makes this class of message highly useful in supporting the development of situational awareness.

Tweets were inspected by category (see code below) to classify the general use-case of each source and eliminate those which were only used for automated/unrelated purposes:

In [28]:
# Investigate Tweets by source to characterise source content:
# (samples are not limited to data_source>0)

ts = Tweet.objects.filter(data_source__gte=1)
fieldname = 'source'
counts = ts.values(fieldname).order_by(fieldname).annotate(count=Count(fieldname)).order_by('-count')
source_list = [x['source'] for x in counts if x['count'] > 10]

for source in source_list:
    print('__________________\n{}:\n'.format(source))
    ts = Tweet.objects.filter(source=source)
    for t in ts.order_by('?')[:15]:
#     for t in ts.filter(data_source__gte=1).order_by('?')[:15]: # To only sample detected Tweets
        print(t.text.replace('\n', ''))

__________________
Instagram:

SO MUCH LOVE FROM SODAK! My heart is swelling! I'm safe and… https://t.co/57NEdX2zgo
Finished colored Capricorn illustration🌊🐏 https://t.co/PFX7YJQ0i7
Helping an Energy family recover after the storm. We may not be at school, but we are definitely… https://t.co/hWCzqXnPo9
CHECK OUT THIS EXPLICT SINGLE "NASTY BISH"... KILLED IT #LA #LOUISVILLE #ableton #cubase #edm… https://t.co/Tmw8VqpASb
#cajunnavy #katytx #hurricaneharvey #houston #htown #rescueandsearch #recoverysearch https://t.co/M8pJtR5ZNU
When he looks at me and tell me none of those girlsmean nothing it's you who I want I'm done ✅… https://t.co/vrNeTHq34y
#KatyPerry on the red carpet at the #VMAs Posted by theoriginaldatvegasgyrl https://t.co/f7XscM9L8n
#foodporn #chicken #mushrooms  #marsala Delicious dinner with no gram Nina No gram Tom No gram… https://t.co/ukYfj8N7AF
This is why I'm only sharing personally confirmed information of people in need of assistance.… https://t.co/X3NhNuljs7
I'm on t

La cuarta temporada de #Gotham incluirá a #Batman https://t.co/QO9bNXy03z
Antes de ser famosa, @shakira ganó un concurso por el trasero más grande https://t.co/o8cC3iJLWl
Tighten your #CORE with these 6-Minute #Yoga moves! #fitness #healthy #wellnesswednesday #meditation https://t.co/JgS5xLEo13 https://t.co/FEVPomIXus
When you're ready to call it quits, some things to consider ... https://t.co/NZIM6G85qd
NOT SO DEAD by @charlielevin #FREE TODAY! Featured on The Top 100 Best Free Kindle Books on « @DigitalBkToday https://t.co/ysdO34hKut
@AndyRichter "...and he cast his gaze across miles of ocean, with nary a drop to drink..."
Good afternoon! Sunny, cool &amp; breezy conditions across #SNE with temperatures currently in the 50's/60's: https://t.co/ZBrYNaprD7
Thalia celebra sus 46 años con un sexy bikini @thalia  #Thalia #Espectaculos https://t.co/Utu4jAm3WK https://t.co/uPmO3cGw9B
#EHArchivo Le traemos lo mejor de @Carmilla_Wyler ► https://t.co/8yf7p8PyHa  https://t.co/fXcCuUHHug
https:/

RT @Steven_Hyden: Biggest "actually" in history. https://t.co/OR2DpjwsA0
Where’s Your Wife Wednesday? 6-9p.m. Enjoy $10 OFF EVERY bottle of wine!#TheWoodlands #WineBar #PatioSeating #BoutiqueWine #LadiesSpecial https://t.co/YznVzGHJSk
https://t.co/m261zAXgv7 #punchproud #follow #salford #liverpool #preston #GBBOGREAT BEDS BEST OFFERS #retweet chance to #WinVisit Website https://t.co/ew8Xu3mHQr
Big B to host 9th edition of KBC onSony from tom@SrBachchan@KBCsony@SonyTV@SonyTv_FC#KaunBanegaCrorepati#kbc9https://t.co/WJBl3uPYKd https://t.co/9dwQttH6Ze
RT @INRedCross: Shout out to @KevinHart4real for helping the #RedCross raise funds for #HurricaineHarvey! Way to advance the humanitarian m…
Do you recognize this woman? @HooverPD say she shoplifted $736.00 worth of merchandise 08/16/2017 at CVS Chace Dr.  $736.00. Tattoo on back? https://t.co/uTPIJLT8MN
RT @GeorgeStrait: We are working on putting together relief efforts with the whole country music community. – GS
RT @Ryan_Wichman: For those

Drinking a Rodeo Clown Double IPA by @karbachbrewing - https://t.co/plrZ5KLFAR
I just earned the 'Hopped Down (Level 26)' badge on @untappd! https://t.co/gYaCrNsb3v
Drinking a Summer's Wit by @Buffbrew - https://t.co/uyTZQAnk5w
Ahh, childhood pie in Salida! - Drinking a Strawberry Rhubarb by @greatdividebrew - https://t.co/BAL6hJYqdI #photo
I just earned the 'Hopped Down' badge on @untappd! https://t.co/bmev2Ng4sl
#hurricaneharvey - Drinking a Houston Haze by @SpindleTap @ Montrose  — https://t.co/kx5G8914SS #photo
I just earned the 'Verified Adventure (Level 100)' badge on @untappd! https://t.co/BlcWD7d3Zo
Dustin's pie shots are the tits! - Drinking a Category 3 Tripel by @hollerbeer - https://t.co/JXvdRBYFya #photo
Needs age. - Drinking a Mash &amp; Coconut (2017) by @thebruery at @thegrowlerbar  — https://t.co/zOWguVrVIT #photo
One of my faves in the MV line up. - Drinking a Scots Gone Wild (Mysterium Verum) (2017) @ EaDo - https://t.co/Z0kxjQHVVn #photo
Juicy AF! Thanks Hop R. For 

@martin_compston Please beat PSG in CL.
Is Arena going to make an adjustment? #USMNT
RT @USCG: People in distress from #harvey can call the following #USCG numbers:281-464-4851281-464-4852281-464-4853281-464-4854281-46…
Rain is cold and wind is strong out here on I-10 and Highway 6#Houston
RT @BillBishopKHOU: This is my neighborhood.  My home has flooded.  That's my wonderful son.  This was his birthday weekend.  I love you. #…
@JoshMom96 @JeniCamCakes This isn’t about reward. This is publicly denouncing hypocrisy/greed among supposed people of faith. Exactly what Jesus did. What reward?
@parallelcross @lgillen 🤣
Whoa. https://t.co/p7Gb3D5keC
Really smart piece —&gt; https://t.co/rBTe7tq9OJ
RT @law4usc: @TxStormChasers Running list of ppl still needing rescue in Houston incl GPS coordinates  https://t.co/FR1rgxJnQi
RT @RealSkipBayless: Turning point: Early in R9, Conor had Floyd in trouble in the corner, stung him, and ref broke it up. Conor argued w r…
RT @CassandraRules: Sarsour is s

This is from a 1962 car insurance commercial showing the number of drivers who could save with State Farm. #Throwback #FromTheArchives https://t.co/3r7NChK5ay
Our hearts go out to everyone impacted by #Harvey. If you would like help those in need, consider donating to the @RedCross.
Five things you should know in time for #backtoschool season https://t.co/JxmqkJRl1u
With stocks entering seasonally weak period it's time to be opportunistic. Read more in The GIC Weekly https://t.co/cumiP7qsOG
Can you use the funds from a 529 college plan for student loans? Find out if you have the right answer. https://t.co/YdM3y9qifu
I hate rushing in the morning. In time for #backtoschool season, how to get your kids out the door https://t.co/kK7B4RO6gX
Trees are down. Power is out. Now what? Allstate customers with damage after #HurricaneHarvey can call 1-800-547-8676 We are here.
We all smile in the same language. Pass some along today. https://t.co/WHarSfgKdn
#NationalDogDay is August 26. Here’s to 

MIT Reveals Tech That Could Prevent Billions in Lost Warehouse Inventory https://t.co/MgdLR5JLmn But more jobs lost to drones &amp; robots.
@donlemon Obama approval ratings. Newsmax CEO was correct Don. low to high 40’s most of his term. Trump at 41 now. https://t.co/hddAgZECFA
Will luxury markets embrace cryptocurrencies before other markets? Will cryptocurrencies become too myriad before development is perfected?
@realDonaldTrump Good! He was only doing his job and liberal pansy judges/DA went on a crusade against him. #withhunt #istandwithtrump #maga
BREAKING: Illinois AG Lisa Madigan suing to force federal court oversight of the Chicago Police Department. https://t.co/yDqgu23hEC... https://t.co/vVM9wyrMXv
7) Comey gets fired all of a sudden mid investigation, Trump goes behind his staff, schedules own interview and basically tells on himself
@wxonmars Oh yes, certainly don’t mean to suggest that they are similar in any way except the regularity of their appearance.
everytime @JoyAn

Could the @backstreetboys be planning an EDM album? Details: https://t.co/n9pGsOPsAY https://t.co/HMxUP1aKZX
How NOT to launch the biggest social campaign of your career! Find out. https://t.co/hOvN8vPYQW #hcsm #MCSMN https://t.co/r0DTw1nO48
Would you like to help with the #HurricaneHarvey #relief effort? Check out these organizations leading the charge: https://t.co/HCvkfg2eL3 https://t.co/Adp6veYeWv
RT @KHOU: FEMA: Over 50 counties in Texas impacted by Harvey.  More right now on KHOU 11 or here: https://t.co/3kDqOMQZ3q https://t.co/IGgA…
Let's go Cards! Thank you @Cardinals for your continued support of healthy moms and babies through the Signature Chefs Auction #MODSTLCHEF https://t.co/Ya8pEFY99E
EXCLUSIVE INTERVIEW: @OldDominion Reveal Their Favorite Songs From New Album, #HappyEndings https://t.co/bLrZ89tsAk https://t.co/4nHzhnBMA5
@crazytallmom Proud to partner with @RedCross to donate 565 cases of products for #Harvey relief efforts https://t.co/ZHzU8qbmMh
@Smilnsinger Sorry to 

China's bond market is predicted to double in size over the next 5 years. How are you seizing this opportunity? https://t.co/sb5SeLMuVR https://t.co/HDKlhtMA4j
The Secret to Being the Smartest Member on Your Association Board https://t.co/pT5wcRnJGd
RT @Associa: NEWS: Associa Begins Cleanup Efforts after #HurricaneHarvey Destruction: https://t.co/vyxodaO7NP
RT @Associa: Associa Cares Earmarks $50k to help families impacted by #Harvey #Associacares https://t.co/njxCpx6mDn
How to provide better, faster, cheaper service for customers in a #DigiWealth world of robo-advisors and big data? https://t.co/Gyvrs1sYQK https://t.co/KhoAGUnONP
RT @Associa: NEWS: Associa Begins Cleanup Efforts after #HurricaneHarvey Destruction: https://t.co/zPhTboCvmL
Find how you can successfully face the challenges of #DigiWealth transformation era and explore our microsite: https://t.co/teBBgTUo7N https://t.co/tA8SLWO4ZG
Do #email blasts the right way with @SendGrid CMO @ScottHeime's 4 tactics @Marketingland: ht

@EmekaGift: #Africa #America #BuhariNigeria BIAFRA: STARVATION, ONE OF BRITAIN'S MAJESTIC RESOURCES IN THE KILLING OF BIAFRANS. https://t...
#Africa #America #Biafra OZUBULU MASSACRE: ALARM ON THE PLANNED ASSASSINATION OF MR ALOYSIUS IKEGWUONU A.K.A... https://t.co/L0W7NoXrrk
Liked “@GrandPrairiePD liked:@GrandPrairiePD @Walmart Timem.twit
#MedbotNews #TwitterRPafterDark: @Dark_Helmet_SB am boinked @SharaTheOrion full of #ButtDiseases.Him almost got trapped in there, too.
@KEENHONORTRUST: .@Jerusalem_Post: #ISRAEL  Declares #Hebron An Official Jewish Settlement! https://t.co/AAJCWN1bco https://t.co/MApGjoKbIl
#TwitterNews #OldFolksNews: #SeniorCitizens in #TwitterPhoto am #rescued from #NursingHome https://t.co/J7D8SUb28E https://t.co/Lnbe0v8iOl
Trump: THANK YOU to all of the great volunteers helping out with #HurricaneHarvey relief in Texas! https://t.co/xzYg12yaTV.
#Africa #America #Biafra: #Nigeria #Is #The #Dwarf #Of Africa https://t.co/ispMMEgZxW https://t.co/DbdPnsiQql...
@EmekaG

@maggieNYT Surprising? This is Trump. It's expected from him.
RT @thehill: GOP lawmaker hits Trump on Arpaio pardon: He pardoned someone who ignored the Bill of Rights https://t.co/orWf7qICIn https://t…
RT @crampell: If you're not disgusted enough: read this on how Arpaio's office botched 400 sex-crime cases, many involving kids  https://t.…
RT @SenatorShaheen: Women have waited long enough @StevenMnuchin1. Urge you to commit to this popular effort that has bipartisan support ht…
RT @PennStateIFC: With the current devastation in Houston, we are pledging $0.15 for every RT this gets! Please forward this along to help…
@johninkingwood https://t.co/vuWG0MW0vJ
A bit ago in Cypress, TX!  #txwx https://t.co/UpVjWUijgW
RT @EricLiptonNYT: Letter cites my NYT story on Icahn from March, as well as more recent New Yorker piece. https://t.co/27fyLg7r6P
@JTownsend96 🤣🤣🤣 so true!
RT @EKillingerNRG: Judge @EdEmmett announced NRG Center will open for #Harvey evacuees  #HoustonFlooding @NRGParkFan #Hou

United Ways in Texas and Louisiana are coordinating #HurricaneHarvey recovery efforts. Please see how you can help. https://t.co/lvr7amYtTY
9 Lessons to Inspire Women Entrepreneurs via @Cornell_Tech: https://t.co/tyB6ifDTd7
Getting by is still a struggle for most, no matter how high the amount on their paycheck. https://t.co/C3Pm54Ohgw
RT @YaniAtCVS: DYK that in addition to @CVSHealth donation, the company is matching employee donations up to $25k! #TeamCVS #Harveyhttps:/…
How #CornellTech researchers are using Instagram to study world cultural trends: https://t.co/x5haiM3iJw
Help us support communities affected by #HurricaneHarvey. https://t.co/1VNt3oUxfx
Here's how you can help support communities and families affected by #HurricaneHarvey. https://t.co/3KYbNtgRUD
A brief history of Roosevelt Island and how #CornellTech will fit into its tradition of community https://t.co/n0DhkEcEB4
A collection of stand-out moments showing the humanity in the face of this unfathomable disaster. (via

* Instagram -- Crossposting images
* Twitter for iPhone
* Twitter Web Client
* Twitter for Android
* Paper.li -- Automated, seems like spam from a single source
* Hootsuite -- Suite, but appears to be used by 'regular users'
* TweetMyJOBS -- Automated job listings
* SafeTweet by TweetMyJOBS -- Automated job listings
* IFTTT -- Appears to be commonly used for cross-posting from eg Instagram
* Facebook
* TweetDeck
* Twitter for iPad
* BubbleLife -- Suite, news
* Twitter Lite
* Buffer -- Suite, news
* Foursquare -- Crossposting check-in app, sometimes informative, but mostly "I'm at x" template
* Untappd -- Crossposting check-in app (for beer drinkers), rarely informative
* SocialNewsDesk -- Suite, news
* Sprout Social -- Suite, news
* Error-log -- Automated bot / spam
* Botize -- All in Spanish, appeared to be only news
* Tweetbot for iΟS -- third-party app, regular users
* Periscope -- Crossposting, live video feeds
* VoiceStorm -- Suite, CRM
* Twitter for Windows
* Google -- Crossposting, mostly from Youtube (user likes or posts a video)
* Hearsay Social -- Suite, news
* despa ringtones -- Automated bot / spam
* circlepix -- Automated retail listings
* TweetCaster for Android -- Third party app
* TownTweet -- Automated forecasts
* Twitter for Mac
* Mobile Web (M2) -- (mobile.twitter.com)
* HubSpot -- Suite, CRM
* Crowdfire - Go Big -- Suite, news
* iOS
* Sprinklr -- Suite, CRM
* LinkedIn -- News article links
* Echofon -- Third-party app, regular users
* SeeYourWeather.com Galveston -- Automated weather bot
* WordPress.com -- Automated blog reposting
* GaggleAMP -- Suite, recruitment
* Twitter for Windows Phone
* Twitter for Android Tablets
* Radian6 -Social Media Management -- Automated news
* Convey: Make it post for you -- Automated news
* Cloudhopper -- Appears to be a Suite, CRM
* dlvr.it -- Automated news
* Twitter for BlackBerry
* Squarespace -- Automated blog reposting
* SocialFlow -- Suite, news
* Donate a Photo -- Looks like some J&J sponsored app
* Twitterrific -- Third-party app, regular users
* a7aa7o -- Automated bot / spam
* CoSchedule -- Suite, CRM
* despacitoapp -- Automated bot / spam
* shortage bot -- Automated bot / spam
* Bambu by Sprout Social -- Suite, CRM
* SocialPilot.co -- Suite, CRM
* Weather Message -- Automated weather reports
* RadarScope -- Weather enthusiast app


Therefore a list of sources can be created which are not used by 'regular users':

We can now compare the rate of useful messages for each source categorisation:

In [29]:
fp = ['Twitter for iPhone',
 'Twitter Web Client',
 'Twitter for Android',
 'TweetDeck',
 'Twitter for iPad',
 'Twitter Lite',
 'Twitter for Windows',
 'Twitter for Mac',
 'Mobile Web (M2)',
 'iOS',
 'Twitter for Windows Phone',
 'Twitter for Android Tablets',
 'Twitter for BlackBerry']

tp = [
 'Tweetbot for iΟS',
 'TweetCaster for Android',
 'Echofon',
 'Twitterrific']
      
cross = ['Instagram',
 'IFTTT',
 'Facebook',
 'Foursquare',
 'Untappd',
 'Periscope',
 'Google',
 'LinkedIn',
 'WordPress.com',
 'Squarespace',
 'RadarScope']

cross2 = ['IFTTT',
 'Facebook',
 'Foursquare',
 'Untappd',
 'Periscope',
 'Google',
 'LinkedIn',
 'WordPress.com',
 'Squarespace',
 'RadarScope']
      
suites = [
 'Hootsuite',
 'BubbleLife',
 'Buffer',
 'SocialNewsDesk',
 'Sprout Social',
 'Botize',
 'VoiceStorm',
 'Hearsay Social',
 'HubSpot',
 'Crowdfire - Go Big',
 'Sprinklr',
 'GaggleAMP',
 'Radian6 -Social Media Management',
 'Convey: Make it post for you',
 'Cloudhopper',
 'SocialFlow',
 'CoSchedule',
 'Bambu by Sprout Social',
 'SocialPilot.co']
      
auto = [
 'TweetMyJOBS',
 'SafeTweet by TweetMyJOBS',
 'circlepix',
 'TownTweet',
 'SeeYourWeather.com Galveston',
 'dlvr.it',
 'Donate a Photo',
 'a7aa7o',
 'Weather Message']
      
spam = ['Paper.li',
 'Error-log',
 'despa ringtones',
 'shortage bot',
 'despacitoapp']

source_cats = {'First Party': fp, 'Third Party': tp, 'Crossposting': cross,
               'Crossposting excl. Insta': cross2, 'Suites': suites, 
               'Automated': auto, 'Spam': spam}

tweet_codings = (Coding.objects
                    .filter(coding_id=1)
                    .filter(tweet__isnull=False)
                    .filter(data_code__data_code_id__gt=0)
                  )
ts = Tweet.objects.filter(coding_for_tweet__in=tweet_codings)
tw_df = pd.DataFrame(list(ts.values('tweet_id', 'source', 'data_source')))

desired_codes = ['Aid Request', 'Ground Truth']
#TODO: check why get returned two vals for some here? :
tw_df['is_desired'] = tw_df['tweet_id'].apply(lambda x: True if tweet_codings.filter(tweet__tweet_id=x)[0].data_code.name in desired_codes else False)
desired_series = tw_df['is_desired']

indv_dict = {}
aggr_dict = {}
for cat in source_cats:
    for source in source_cats[cat]:
        s1 = tw_df['source'].apply(lambda x: True if x == source else False)
        pos = sum(s1 & desired_series)
        tot = sum(s1)
        if tot == 0:
            continue
        conf = confusion_matrix_from_series(s1, desired_series)
        agr = calc_agreement_metrics(conf)
        prop = pos/tot*100
        print('{:.1f}% -- {}: {} of {} -- Pr: {:.3f}, R: {:.3f}'.format(
            prop, source, pos, tot, agr['Prec'], agr['Recall']
            ))
        indv_dict[source] = {'Positive': pos, 'Total': tot, 'Precision' : agr['Prec'], 'Recall': agr['Recall']}

    s1 = tw_df['source'].apply(lambda x: True if x in source_cats[cat] else False)
    conf = confusion_matrix_from_series(s1, desired_series)
    agr = calc_agreement_metrics(conf)
    pos = sum(s1 & desired_series)
    tot = sum(s1)
    
    print('{}: {:.2f}% of {}'.format(cat, pos/tot*100, tot))
    print('Precision: {:.3f}, Recall: {:.3f}\n'.format(agr['Prec'], agr['Recall']))
    
    aggr_dict[cat] = {'Positive': pos, 'Total': tot, 'Precision' : agr['Prec'], 'Recall': agr['Recall']}

14.6% -- Twitter for iPhone: 59 of 405 -- Pr: 0.146, R: 0.144
8.0% -- Twitter Web Client: 33 of 415 -- Pr: 0.080, R: 0.081
13.0% -- Twitter for Android: 31 of 238 -- Pr: 0.130, R: 0.076
8.6% -- TweetDeck: 3 of 35 -- Pr: 0.086, R: 0.007
5.7% -- Twitter for iPad: 2 of 35 -- Pr: 0.057, R: 0.005
10.0% -- Twitter Lite: 3 of 30 -- Pr: 0.100, R: 0.007
0.0% -- Twitter for Windows: 0 of 7 -- Pr: 0.000, R: 0.000
0.0% -- Twitter for Mac: 0 of 2 -- Pr: 0.000, R: 0.000
0.0% -- Mobile Web (M2): 0 of 5 -- Pr: 0.000, R: 0.000
66.7% -- Twitter for Windows Phone: 2 of 3 -- Pr: 0.667, R: 0.005
0.0% -- Twitter for BlackBerry: 0 of 2 -- Pr: 0.000, R: 0.000
First Party: 11.30% of 1177
Precision: 0.113, Recall: 0.325

0.0% -- Tweetbot for iΟS: 0 of 1 -- Pr: 0.000, R: 0.000
0.0% -- TweetCaster for Android: 0 of 5 -- Pr: 0.000, R: 0.000
Third Party: 0.00% of 6
Precision: 0.000, Recall: 0.000

33.7% -- Instagram: 247 of 733 -- Pr: 0.337, R: 0.604
17.6% -- IFTTT: 6 of 34 -- Pr: 0.176, R: 0.015
17.1% -- Facebook:

/home/rosles/projects/crisis-data/venv/lib/python3.6/site-packages/ipykernel_launcher.py:84: RuntimeWarning: invalid value encountered in double_scalars


75.0% -- Periscope: 3 of 4 -- Pr: 0.750, R: 0.007
25.0% -- Google: 1 of 4 -- Pr: 0.250, R: 0.002
0.0% -- LinkedIn: 0 of 2 -- Pr: 0.000, R: 0.000
0.0% -- WordPress.com: 0 of 2 -- Pr: 0.000, R: 0.000
Crossposting excl. Insta: 17.48% of 103
Precision: 0.175, Recall: 0.044

10.0% -- Hootsuite: 6 of 60 -- Pr: 0.100, R: 0.015
1.8% -- BubbleLife: 1 of 57 -- Pr: 0.018, R: 0.002
0.0% -- Buffer: 0 of 14 -- Pr: 0.000, R: 0.000
9.1% -- SocialNewsDesk: 1 of 11 -- Pr: 0.091, R: 0.002
0.0% -- Sprout Social: 0 of 3 -- Pr: 0.000, R: 0.000
0.0% -- Botize: 0 of 6 -- Pr: 0.000, R: 0.000
100.0% -- VoiceStorm: 1 of 1 -- Pr: 1.000, R: 0.002
0.0% -- Hearsay Social: 0 of 3 -- Pr: 0.000, R: 0.000
0.0% -- HubSpot: 0 of 1 -- Pr: 0.000, R: 0.000
25.0% -- Crowdfire - Go Big: 1 of 4 -- Pr: 0.250, R: 0.002
0.0% -- GaggleAMP: 0 of 2 -- Pr: 0.000, R: 0.000
0.0% -- Radian6 -Social Media Management: 0 of 1 -- Pr: 0.000, R: 0.000
0.0% -- Cloudhopper: 0 of 2 -- Pr: 0.000, R: 0.000
Suites: 6.06% of 165
Precision: 0.061, Rec

In [40]:
# Re-Printing output for Latex table
insuff_data_threshold = 10
source_cats = {'First Party Apps': fp, 'Third Party Apps': tp, 'Crossposting Apps': cross, 
               'Social Media Suites': suites, 'Automated Apps': auto, 'Spam': spam}

# for cat in source_cats:
#     for source in source_cats[cat]:
#         s1 = tw_df['source'].apply(lambda x: True if x == source else False)
#         tot = sum(s1)
#         if tot < insuff_data_threshold:
#             continue
#         conf = confusion_matrix_from_series(s1, desired_series)
#         agr = calc_agreement_metrics(conf)
#         prop = pos/tot*100
#         print('{} & {} & {:.3f} & {:.3f} \\\\'.format(source, tot, agr['Prec'], agr['Recall']))

row_list = []
for cat in source_cats:
    s1 = tw_df['source'].apply(lambda x: True if x in source_cats[cat] else False)
    conf = confusion_matrix_from_series(s1, desired_series)
    agr = calc_agreement_metrics(conf)
    tot = sum(s1)
#    print('{} & {} & {:.2f} & {:.2f} \\\\'.format(cat, tot, agr['Prec'], agr['Recall']))   
    row_dict = {'Source Classification': cat, 'Total': tot, 'PrecisionRaw': agr['Prec'], 'RecallRaw': agr['Recall']}
    row_list.append(row_dict)

df = pd.DataFrame(row_list)
df['Description'] = ''
df['Precision'] = df['PrecisionRaw'].round(2)
df['Recall'] = df['RecallRaw'].round(2)

#df.loc[df['Total'] < 10, 'Precision'] = 'Insufficient data'
df['Precision'] = df.apply(lambda x: 'Insufficient data (n={})'.format(
    x['Total']) if x['Total'] < insuff_data_threshold else x['Precision'], axis=1)
df.loc[df['Total'] < insuff_data_threshold, 'Recall'] = ''


df['Description'] = \
    ['Applications developed by Twitter -- these represented the most typical sources used to interact with the platform.',
    'Applications developed by third party developers to interact with Twitter in a similar fashion to first-party apps.',
    'Other social media applications with a feature to automatically crosspost user content to Twitter.',
    'Software designed to manage multiple social media accounts for businesses. Typically used by organisations to post news articles or public relations content.',
    'Applications designed to automatically post content such as periodic weather reports.',
    'A secondary classification of automated apps which were designed to generate spam posts. These often included trending hashtags to increase their exposure.'
    ]

df_to_tex(df, cols=['Source Classification', 'Description', 'Precision', 'Recall'], label='ts-src-tax', caption='Twitter Source Taxonomy', 
              width=1.0, row_sep=3, col_widths='X[-2,lm] X[-4] X[1] X[1]')


Writing df to file as tex...


/home/rosles/projects/crisis-data/venv/lib/python3.6/site-packages/ipykernel_launcher.py:84: RuntimeWarning: invalid value encountered in double_scalars


In [31]:
set_tex_var('hrvTwFPPrProp', aggr_dict['First Party']['Precision']*100, 1)
set_tex_var('hrvTwFPRcProp', aggr_dict['First Party']['Recall']*100, 1)
set_tex_var('hrvTwXPostPr', aggr_dict['Crossposting']['Precision'], 3)
set_tex_var('hrvTwXPostRc', aggr_dict['Crossposting']['Recall'], 3)
set_tex_var('hrvTwXPostTot', aggr_dict['Crossposting']['Total'], 3)
set_tex_var('hrvTwInstaTot', indv_dict['Instagram']['Total'], 3)
set_tex_var('hrvTwInstaPr', indv_dict['Instagram']['Precision'], 3)
set_tex_var('hrvTwInstaRc', indv_dict['Instagram']['Recall'], 3)
set_tex_var('hrvTwHootPr', indv_dict['Hootsuite']['Precision'], 3)
set_tex_var('hrvTwHootRc', indv_dict['Hootsuite']['Recall'], 3)
# Suites stats excluding HootSuite:
set_tex_var('hrvTwSMPrExclHoot', sum([indv_dict[x]['Positive'] for x in indv_dict.keys() if (x in suites and x != 'Hootsuite')]) / 
            sum([indv_dict[x]['Total'] for x in indv_dict.keys() if (x in suites and x != 'Hootsuite')]), 3)
set_tex_var('hrvTwSMRcExclHoot', sum([indv_dict[x]['Recall'] for x in indv_dict.keys() if (x in suites and x != 'Hootsuite')]), 3)
set_tex_var('hrvTwAutPos', aggr_dict['Automated']['Positive'], 3)
set_tex_var('hrvTwAutTot', aggr_dict['Automated']['Total'], 3)
set_tex_var('hrvTwSpamPos', aggr_dict['Spam']['Positive'], 3)
set_tex_var('hrvTwSpamTot', aggr_dict['Spam']['Total'], 3)


The sources classified as spam, automated and suites (once Hootsuite is excluded) have neglible or zero rates of useful content and may therefore be eliminated from the dataset. Crossposting apps vary by app and should be classified individually. The sample of third-party apps is too small to draw conclusions.

The list of unwanted sources is then defined:

In [32]:
unwanted_sources = [
    'Paper.li',
    'TweetMyJOBS',
    'SafeTweet by TweetMyJOBS',
    'BubbleLife',
    'Buffer',
#     'Foursquare', ###
    'Untappd', ###
    'SocialNewsDesk',
    'Sprout Social',
    'Error-log',
    'Botize',
    'VoiceStorm',
    'Google', ###
    'Hearsay Social',
    'despa ringtones',
    'circlepix',
    'TownTweet',
    'HubSpot',
    'Crowdfire - Go Big',
    'Sprinklr',
    'LinkedIn',
    'SeeYourWeather.com Galveston',
    'WordPress.com',
    'GaggleAMP',
    'Radian6 -Social Media Management',
    'Convey: Make it post for you',
    'Cloudhopper',
    'dlvr.it',
    'Squarespace',
    'SocialFlow',
    'Donate a Photo',
    'a7aa7o',
    'CoSchedule',
    'despacitoapp',
    'shortage bot',
    'Bambu by Sprout Social',
    'SocialPilot.co',
    'Weather Message',
    'RadarScope']

To verify these classifications, we can compare the content of each source subset using the coded Tweet dataset. Where a significant proportion of the subset has been classified as a 'desired code', the decision to eliminate the source should be reconsidered.

In [33]:
# #from django.db.models import Count
# tweet_codings = (Coding.objects
#                     .filter(coding_id=1)
#                     .filter(tweet__isnull=False)
#                     .filter(data_code__data_code_id__gt=0)
#                   )

# #desired_codes = ['Aid Request', 'Ground Truth', 'Info for Affected', 'Emotion - Affected']
# desired_codes = ['Aid Request', 'Ground Truth']
to_inspect = []

for source in unwanted_sources:
    cs = tweet_codings.filter(tweet__source=source)
    tot = cs.count()
    if tot == 0:
        continue
    fieldname = 'data_code__name'
    counts = cs.values(fieldname).annotate(the_count=Count(fieldname))
    n = 0
    for c in counts:
        if c['data_code__name'] in desired_codes:
            n += c['the_count']
    if tot >= 1:
        print('{:.1f}% -- {}: {} of {}'.format(n/tot*100, source, n, tot))
        # If more than 10% desired, add to list for manual inspection:
        if n/tot > .1:
            to_inspect.append(source) 

0.0% -- Paper.li: 0 of 74
0.0% -- TweetMyJOBS: 0 of 90
0.0% -- SafeTweet by TweetMyJOBS: 0 of 101
1.8% -- BubbleLife: 1 of 57
0.0% -- Buffer: 0 of 14
0.0% -- Untappd: 0 of 13
9.1% -- SocialNewsDesk: 1 of 11
0.0% -- Sprout Social: 0 of 3
0.0% -- Error-log: 0 of 8
0.0% -- Botize: 0 of 6
100.0% -- VoiceStorm: 1 of 1
25.0% -- Google: 1 of 4
0.0% -- Hearsay Social: 0 of 3
0.0% -- despa ringtones: 0 of 5
0.0% -- circlepix: 0 of 2
0.0% -- TownTweet: 0 of 8
0.0% -- HubSpot: 0 of 1
25.0% -- Crowdfire - Go Big: 1 of 4
0.0% -- LinkedIn: 0 of 2
0.0% -- SeeYourWeather.com Galveston: 0 of 1
0.0% -- WordPress.com: 0 of 2
0.0% -- GaggleAMP: 0 of 2
0.0% -- Radian6 -Social Media Management: 0 of 1
0.0% -- Cloudhopper: 0 of 2
0.0% -- Donate a Photo: 0 of 1
0.0% -- Weather Message: 0 of 1


In [34]:
# Manually inspect the Tweets from the sources with >10% useful codes:

for source in to_inspect:
    cs = tweet_codings.filter(tweet__source=source)
    cs2 = cs.filter(data_code__name__in=desired_codes)
    print('\n{}: {} of {}'.format(source, cs2.count(), cs.count()))
    for c in cs2:
        print(c.tweet.text)


VoiceStorm: 1 of 1
#Harvey is strong, but hearts are stronger. https://t.co/te1uSYYtdp https://t.co/DgMCKaWwvN

Google: 1 of 4
I added a video to a @YouTube playlist https://t.co/cnFgr5WDZa Aerial view: Texas Flooding Aftermath [part 3] #Houston #TXwx #Harvey

Crowdfire - Go Big: 1 of 4
MỚI NHẤT** HOUSTON** , TEXAS** CHÌM TRONG BIỂN NƯỚC SAU SIÊU BÃO **HARVEY** - YouTube
https://t.co/GlnCiqwIV8 https://t.co/JSZgmmxZEK


The qualitative inspection of the sources with a significant proportion of useful information shows that the positively coded Tweets are not strongly associated with their classification, and as single instances, are only flagged due to the low sample size. Therefore these discrepencies may be ignored.

Now we can check how much of the dataset was made up of the uninformative sources and thus measure the impact of early elimination of these data:

In [35]:
ts = Tweet.objects.filter(data_source__gt=1)
print('Total for data_source=3 (geo stream)', ts.count())
ts2 = ts.exclude(source__in=unwanted_sources)
print('Total from accepted sources', ts2.count())
print('Total removed: {} ({:.2f}%)\n'.format((ts.count() - ts2.count()), 
                                           ((ts.count() - ts2.count())/ts.count())*100))

print('Accepted-source adjusted source distribution:')
fieldname = 'source'
counts = ts2.values(fieldname).order_by(fieldname).annotate(count=Count(fieldname)).order_by('-count')

for x in [c for c in counts if c['count'] > 10]:
    print('{:.1f}% {}: {}'.format((x['count']/ts2.count()*100), x['source'], x['count']))

Total for data_source=3 (geo stream) 15569
Total from accepted sources 12628
Total removed: 2941 (18.89%)

Accepted-source adjusted source distribution:
94.4% Instagram: 11922
2.1% Foursquare: 267
1.0% Twitter for Android: 129
0.8% Hootsuite: 104
0.7% Twitter for iPhone: 87
0.2% iOS: 28
0.2% Twitter for Android Tablets: 27
0.1% Twitter for Windows Phone: 15
0.1% Tweetbot for iΟS: 12


In [36]:
set_tex_var('hrvTwGeoIrr', (ts.count() - ts2.count()))
set_tex_var('hrvTwGeoIrrProp', (ts.count() - ts2.count())/ts.count()*100, 1)

In [37]:
ts = Tweet.objects.filter(data_source=1)
print('Total for data_source=1 (keyword stream)', ts.count())
ts2 = ts.exclude(source__in=unwanted_sources)
print('Total from accepted sources', ts2.count())
print('Total removed: {} ({:.2f}%)\n'.format((ts.count() - ts2.count()), 
                                           ((ts.count() - ts2.count())/ts.count())*100))

print('Accepted-source adjusted source distribution:')
fieldname = 'source'
counts = ts2.values(fieldname).order_by(fieldname).annotate(count=Count(fieldname)).order_by('-count')

for x in [c for c in counts if c['count'] > 10]:
    print('{:.1f}% {}: {}'.format((x['count']/ts2.count()*100), x['source'], x['count']))

Total for data_source=1 (keyword stream) 31303
Total from accepted sources 28081
Total removed: 3222 (10.29%)

Accepted-source adjusted source distribution:
31.9% Twitter for iPhone: 8959
28.4% Twitter Web Client: 7989
18.1% Twitter for Android: 5086
3.5% Hootsuite: 981
2.6% Instagram: 733
2.6% IFTTT: 720
2.4% Facebook: 682
2.3% TweetDeck: 658
2.3% Twitter for iPad: 652
2.1% Twitter Lite: 595
0.2% Periscope: 70
0.2% Twitter for Windows: 62
0.2% Tweetbot for iΟS: 59
0.2% Twitter for Mac: 51
0.2% TweetCaster for Android: 51
0.2% Mobile Web (M2): 50
0.1% Echofon: 40
0.1% Twitter for BlackBerry: 19
0.1% Twitter for Windows Phone: 16
0.1% iOS: 15
0.0% Twitterrific: 14


In [38]:
set_tex_var('hrvTwKwTotal', ts.count())
set_tex_var('hrvTwKwIrr', (ts.count() - ts2.count()))
set_tex_var('hrvTwKwIrrProp', (ts.count() - ts2.count())/ts.count()*100, 1)

We can test the impact of removing the 'undesired' sources on the coded dataset, which shows improvements in 'precision' of the subsets at a very small cost to recall (2-3 relevant messages dropped).

In [48]:
val_dict = {}

for i in [1,3]:
    print('\nData Source:', i)
    a = tw_df.loc[tw_df['data_source'] == i]['is_desired']
    b = tw_df.loc[(tw_df['data_source'] == i) & (~tw_df['source'].isin(unwanted_sources))]['is_desired']
    print(a.value_counts(), a.shape, sum(a.value_counts()))
    print(b.value_counts(), b.shape)

#     print('Original precision: {:.3f}, Filtered precision: {:.3f}'.format(a.value_counts()[1]/sum(a.value_counts()), 
#                                                                 b.value_counts()[1]/sum(b.value_counts()) ))
#     print('Original recall: {:.3f}, Filtered recall: {:.3f}'.format(a.value_counts()[1]/a.value_counts()[1], 
#                                                                 b.value_counts()[1]/a.value_counts()[1] ))
    row_name = {1:'Keyword', 3:'Geographic'}
    val_dict[row_name[i]] = {'Base Precision': a.value_counts()[1]/sum(a.value_counts()),
                             'Filtered Precision': b.value_counts()[1]/sum(b.value_counts()),
                             'Base Recall': 1.0, 
                             'Filtered Recall': b.value_counts()[1]/a.value_counts()[1]}
    
print('\nCombined')
a = tw_df.loc[tw_df['data_source'] > 0]['is_desired']
b = tw_df.loc[(tw_df['data_source'] > 0) & (~tw_df['source'].isin(unwanted_sources))]['is_desired']
print(a.value_counts(), a.shape, sum(a.value_counts()))
print(b.value_counts(), b.shape)

# print('Original precision: {:.3f}, Filtered precision: {:.3f}'.format(a.value_counts()[1]/sum(a.value_counts()), 
#                                                             b.value_counts()[1]/sum(b.value_counts()) ))
# print('Original recall: {:.3f}, Filtered recall: {:.3f}'.format(a.value_counts()[1]/a.value_counts()[1], 
#                                                             b.value_counts()[1]/a.value_counts()[1] ))

val_dict['Combined'] = {'Base Precision': a.value_counts()[1]/sum(a.value_counts()),
                         'Filtered Precision': b.value_counts()[1]/sum(b.value_counts()),
                         'Base Recall': 1.0, 
                         'Filtered Recall': b.value_counts()[1]/a.value_counts()[1]}


df2 = df.from_dict(val_dict, orient='index')
df2['Stream'] = df2.index
df2 = df2.iloc[::-1]
df2 = df2[['Stream'] + df2.columns.tolist()[:-1]]
df2 = df2.round(3)


df_to_tex(df2, cols=['Stream', 'Base Precision', 'Filtered Precision', 'Filtered Recall'], 
          label='src-filtered', caption='Source Filtered Relevancy Measures',
         row_sep=0, dec_dict={'Base Precision':3, 'Filtered Precision':3, 'Filtered Recall':3})

df2


Data Source: 1
False    1343
True      166
Name: is_desired, dtype: int64 (1509,) 1509
False    1201
True      163
Name: is_desired, dtype: int64 (1364,)

Data Source: 3
False    748
True     243
Name: is_desired, dtype: int64 (991,) 991
False    478
True     241
Name: is_desired, dtype: int64 (719,)

Combined
False    2091
True      409
Name: is_desired, dtype: int64 (2500,) 2500
False    1679
True      404
Name: is_desired, dtype: int64 (2083,)
Writing df to file as tex...


Stream  Base Precision  Filtered Precision  Base Recall  \
Keyword        Keyword           0.110               0.120          1.0   
Geographic  Geographic           0.245               0.335          1.0   
Combined      Combined           0.164               0.194          1.0   

            Filtered Recall  
Keyword               0.982  
Geographic            0.992  
Combined              0.988

## TODO

In [ ]:
##TODO:
# Try ML with and without network metrics
# Test at different time periods
# Test other event datasets
# Check steps from book - change to best practice sklearn


# What correlates with witness label
# GPS accounts tend to be spam/businesses?
# Compare GPS from stream / profile / hand coding
# Explore age of account
# Is detecting co-occuring tags viable?
# What kind of data/user is likely to be deleted?
# Is user name change / user deletion/protection a useful predictor
# Compare Change in network, whether it's useful to collect.
# Check gps count, location in profile
# Check timezone distribution
# 'Ordinary person' vs bot/celeb/business/news -- using source field, tweet rate, timezone

# prop of gps -- users and tweets. Automated, instagram sourced?
# prop of sources
# prop of media/urls
# users with location on profile? Some set 'in solidarity'?
# Cycadian posting rythym - can identify real people vs bots?
# location via friend network?
# language

# \item Tweets which were automatically generated from Instagram posts were much more likely to include GPS coordinates, and as media, more likely to represent a ground truth. Therefore this content may be worth focusing on.
# \item Aid requests were very rare. Those that were identified were often reposts rather than originals, and are often referring to the same original message which begins to trend.
# \item Info for affected class should differentiate between immediate and non-immediate content. E.g. a call to mobilise a clean-up or rescue crew vs. a link to an insurance claim form.
# \item For `unrelated' messages, those which matched the keyword stream were highly represented by automated messages coming from a particular set of sources which presumably uses trending tags to gain exposure. This is easy to pre-filter.
# \item Geographically-tagged Tweets are predominantly either: Instagram cross-posts, or automatically generated job listings from a small set of sources (and therefore easy to pre-filter).
        

# Sum of network edge reciprocity
# k-cohesiveness -- Structural cohesion